File: bindings/php/examples/custom-operator.php
<?php

declare(strict_types=1);

// custom-operator: register a PHP `double` operator and call it from a rule.
// Custom operators receive their pre-evaluated arguments as a JSON-array
// string and return a JSON-value string. Built-in names always win.
//
// Run from bindings/php/ (needs `composer install` once, plus the C ABI cdylib:
// `cargo build --release --manifest-path ../c/Cargo.toml`):
//   php examples/custom-operator.php

use Goplasmatic\Datalogic\Engine;

require __DIR__ . '/../vendor/autoload.php';

$engine = Engine::builder()
    ->addOperator('double', function (string $argsJson): string {
        $args = json_decode($argsJson, true);
        return json_encode($args[0] * 2);
    })
    ->build();

echo $engine->apply('{"double": [21]}', '{}'), PHP_EOL; // 42

File: bindings/dotnet/examples/CustomOperator.cs
// custom-operator: register a C# `double` operator and call it from a rule.
// Custom operators receive their pre-evaluated arguments as a JSON-array
// string and return a JSON-value string. Built-in names always win.
//
// Run from bindings/dotnet/:
//   dotnet run --project examples -- custom-operator

using System.Globalization;
using System.Text.Json.Nodes;

namespace Goplasmatic.Datalogic.Examples;

internal static class CustomOperatorExample
{
    internal static void Run()
    {
        using var engine = Engine.Builder()
            .AddOperator("double", argsJson =>
            {
                var n = JsonNode.Parse(argsJson)![0]!.GetValue<double>();
                return (n * 2).ToString(CultureInfo.InvariantCulture);
            })
            .Build();

        Console.WriteLine(engine.Apply("""{"double": [21]}""", "{}")); // 42
    }
}
